Add-404-page

ozer619 opened the pull request:
Minimalist 404 page .

RedFox0x20 commented:
Replaced by #183

RedFox0x20 commented:
Whoops sorry, wrong place

-- These changes look good, worth doing a comparison with #144 to see if any other information may be worth adding.

marekrozmus commented on pages/404.js:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+import ImageGif404 from '../public/images/404.gif';
+export default function NotFound() {
```
please add new line after import

marekrozmus commented on pages/404.js:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <html>
+      <head>
+        <title> {'404 Not Found'} </title>
```
remove the extra spaces before `{` and after `}` - what IDE are you using? If Visual Studio code than you can install `Prettier` extension and set to format code on saving file.

marekrozmus commented on pages/404.js:
```diff
@@ -0,0 +1,31 @@
+            <div className="">
+              <div className="flex justify-center items-center">
+                <div className="lg:w-1/2 pr-4 pl-4">
+                  <Image alt="gif" src={ImageGif404} />
```
this `alt` does not add any useful information

marekrozmus reviewed (changes requested):
General comment - should we use GIF file here? - it is quite big.

audreyfeldroy reviewed (changes requested):
Haha, it's funny and the gif is cute! Thank you for your thoughtful work here @ozer619.

@marekrozmus yes, 1.25mb is large for an image on a web page. 

@ozer619 perhaps try one of these lightweight SVG animations instead? https://www.pixeltrue.com/free-packs/error-state

audreyfeldroy commented:
Further 404 page inspiration: https://octopusenergy.com/404 😃

audreyfeldroy commented:
I've added tips for minimizing file sizes in [Contributor Power Tips](https://margaritahumanitarian.gitbook.io/help-a-family-manual/contributor-power-tips#minimize-file-sizes). Hope this helps 🙂

audreyfeldroy commented on pages/404.js:
```diff
@@ -0,0 +1,43 @@
+            <div className="">
+              <div className="flex justify-center items-center">
+                <div className="lg:w-1/2 pr-4 pl-4">
+                  <Image alt="404-gif" src={ImageGif404} />
```
```suggestion
                  <Image alt="Caveman accidentally getting electrocuted" src={ImageGif404} />
```

audreyfeldroy reviewed (changes requested):
@ozer619 I've made a suggestion for the image alt text.

audreyfeldroy commented on pages/404.js:
```diff
@@ -0,0 +1,43 @@
+                  {
+                    'The page you requested was not found. Always act in kindness by reporting errors like this '
+                  }
+                  <Link href="https://github.com/margaritahumanitarian/helpafamily/issues/new/choose">
```
```suggestion
                  <Link href="https://github.com/margaritahumanitarian/helpafamily/issues/new/choose" passHref>
```

audreyfeldroy commented:
@ozer619 Kindly look into this:

```
32:24  Error: Replace `·href="https://github.com/margaritahumanitarian/helpafamily/issues/new/choose"·passHref` with `⏎····················href="https://github.com/margaritahumanitarian/helpafamily/issues/new/choose"⏎····················passHref⏎··················`  prettier/prettier
--
```

audreyfeldroy commented on pages/404.js:
```diff
@@ -0,0 +1,43 @@
+                    'The page you requested was not found. Always act in kindness by reporting errors like this '
+                  }
+                  <Link href="https://github.com/margaritahumanitarian/helpafamily/issues/new/choose" passHref>
+                    <span className="text-blue-700">{'here'}</span>
```
```suggestion
                    <span className="text-blue-700">{'here.'}</span>
```

RedFox0x20 commented on pages/404.js:
```diff
@@ -0,0 +1,49 @@
+              </div>
+              <div>
+                <Link href="/">
+                  <button class="bg-transparent mt-10 hover:bg-blue-500 text-blue-700 font-semibold hover:text-white py-2 px-4 border border-blue-500 hover:border-transparent rounded">
```
`class` should be `className`

RedFox0x20 replied:
The button should also have `type="button"`

RedFox0x20 reviewed (changes requested):
Changes required as discussed in Discord

`<Link>` elements need the `passHref` attribute and linting errors.
See additional comments.

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,24 @@
+
+    return (
+      <div>
+        <Lottie options={defaultOptions} height={400} width={400} />
```
Properties also need sorting alphabetically.
height, options, width.

RedFox0x20 replied:
Properties need sorting alphabetically.
height, options, width.

RedFox0x20 commented:
^ Wasn't meant to be an approve, hotkeys got messy haha.

RedFox0x20 reviewed (approved):
Nicely done, there are other possible changes such as to styling however these can be left to be done over other iterations with others unless there are any other changes there that you would like to make @ozer619

Nice working with you on this one to build up to this functional 404 page that we have here.

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,24 @@
+
+class AstronautLottie extends Component {
+  render() {
+    const defaultOptions = {
+      loop: true,
+      autoplay: true,
+      animationData: animationData,
+      rendererSettings: {
+        preserveAspectRatio: 'xMidYMid slice',
+      },
+    };
```
Const data can be kept outside of the object, this is an optimisation as it prevents the data being duplicated for every use.

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,24 @@
+      },
+    };
+
+    return (
+      <div>
+        <Lottie height={400} options={defaultOptions} width={400} />
+      </div>
+    );
```
I believe this could be simplified to
```js
return <Lottie height={400} options={defaultOptions} width={400} />
```

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,24 @@
+import Lottie from 'react-lottie';
+import animationData from '../ lotties/astronaut.json';
+
+class AstronautLottie extends Component {
+  render() {
```
Rather than using a class here we should use the functional version:
```
export default function AstronautLottie()
```
As a part of this we can remove the render() and just use the return as shown with other components.

pydanny replied:
@ozer619 Hey, let us know if you need any help with figuring out this change. 😃

RedFox0x20 reviewed (changes requested):
Just some additional changes that I've thought of while looking back through, mostly optimisation and refactoring.

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,18 @@
+import Lottie from 'react-lottie';
+import React from 'react';
+import animationData from '../ lotties/astronaut.json';
+
```
Nicely done converting this back into a functional component as opposed to the class used previously.

Thanks for correcting the spelling error too! However,j ust a note that the file name is still misspelt, it is currently `astronaout.jsx` instead of `astronaut.jsx`.

ozer619 replied:
I corrected the spelling of astronaut and also fixed the space typo :)

RedFox0x20 commented on components/Astronaout.jsx:
```diff
@@ -0,0 +1,18 @@
+  },
+};
+
+function AstronautLottie() {
+  return <Lottie height={400} options={defaultOptions} width={400} />;
+}
+
+export default AstronautLottie;
```
We can simplify this further by combining lines 18 and 14 by doing the following

```suggestion
export default function AstronautLottie() {
  return <Lottie height={400} options={defaultOptions} width={400} />;
}
```

RedFox0x20 reviewed (changes requested):
Nicely done @ozer619 ! Just another couple minor changes to make as far as I can see. Thanks for the work going into this page, has been a good challenge with a few _new_ concepts for us to think about ((Lotties and how they work etc.))

audreyfeldroy commented:
@RedFox0x20 Hey, just wanted to say your reviews on this are fantastic. I'm looking at this with @ozer619 and @Sahil-101 over Discord livestream now and helping them get this over the finish line 😄

ozer619 commented:
Hi @audreyfeldroy @marekrozmus @RedFox0x20  thank you soooo much for reviewing my pull request. I have addresed all requested changes :
- A link/button that takes the user back to the index page /
- Converted Gif to small svg
- Fixed typos
- Changed class component to function component
- Code optimizations  
    - fixed spaces
    - moved const data out of the object
- Used lottie library to work with svg animations
- Link element has passHref
